lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -0,0 +1,481 @@
+                        .values()
+                        .max()
+                        .copied()
+                        .unwrap_or(0)
+                    * VECTOR_ELEMENT_SIZE;
+
+                let segment_config = read_segment.config();
+
+                if read_segment.segment_type() == SegmentType::Special {
+                    return None; // Never optimize already optimized segment
+                }
+
+                // Determine whether segment has mismatch
+                let has_mismatch =
+                    segment_config
+                        .vector_data
+                        .iter()
+                        .any(|(vector_name, vector_data)| {
+                            // Check HNSW mismatch
+                            match &vector_data.index {
+                                Indexes::Plain {} => {}
+                                Indexes::Hnsw(effective_hnsw) => {
+                                    // Select vector specific target HNSW config
+                                    let target_hnsw_collection = &self.hnsw_config;
+                                    let target_hnsw_vector = self
+                                        .collection_params
+                                        .vectors
+                                        .get_params(vector_name)
+                                        .and_then(|vector_params| vector_params.hnsw_config)
+                                        .map(|vector_hnsw| vector_hnsw.update(target_hnsw_collection))
+                                        .and_then(|hnsw| match hnsw {
+                                            Ok(hnsw) => Some(hnsw),
+                                            Err(err) => {
+                                                log::warn!("Failed to merge collection and vector HNSW config, ignoring: {err}");
+                                                None
+                                            }
+                                        });
+                                    let target_hnsw = target_hnsw_vector
+                                        .as_ref()
+                                        .unwrap_or(target_hnsw_collection);
+
+                                    // Select segment if we have an HNSW mismatch that requires rebuild
+                                    if effective_hnsw.mismatch_requires_rebuild(target_hnsw) {
+                                        return true;
+                                    }
+                                }
+                            }
+
+                            false
+                        });
+
+                has_mismatch.then_some((*idx, vector_size))
+            })
+            .collect();
+
+        // Select segment with largest vector size
+        candidates
+            .into_iter()
+            .max_by_key(|(_, vector_size)| *vector_size)
+            .map(|(segment_id, _)| segment_id)
+            .into_iter()
+            .collect()
+    }
```

Comment:
Not a huge fan of this method. It's kinda hard to read with huge nested `filter_map` and `any` calls. I guess, one way to improve it would be to extract parts of it into a few freestanding functions?